### PR TITLE
refactor(buildenv): replace fixed-count retry loop with two-sided path-info termination

### LINKED ISSUE #4348: refactor(buildenv): remove fixed-count retry loop and error-classification machinery from materialise_with_retry
## Background

PR flox/flox#4282 introduced `materialise_with_retry` to handle GC races during environment builds using a `stat()` + `nix path-info` verification approach. However, it also carried over a fixed-count retry loop (`MAX_RETRIES = 3`), exponential backoff, and subsequent PRs ([#4323](<https://github.com/flox/flox/issues/4323>), [#4324](<https://github.com/flox/flox/issues/4324>)) added error classification (`is_transient()`) on the assumption that arbitrary retries with backoff were needed to handle transient failures.

After further analysis with @dcarley, this machinery was based on a misdiagnosis. The actual root cause identified in flox/flox#4327 is that Nix store DB entries can lag behind the filesystem: a store path passes `stat()` but `builtins.storePath` inside `buildenv.nix` fails because the path is not yet registered in the Nix daemon's SQLite database. Related issues flox/flox#4302, flox/flox#4303, flox/flox#4304, flox/flox#4305, and flox/flox#4336 have been closed as superseded.

## Goal

Simplify `materialise_with_retry` to reflect the real failure model:

1. **Optimistic** `stat()` **on the first attempt.** Call `buildenv.nix` after confirming paths are present on the filesystem — no `nix path-info` overhead on the happy path.
2. `nix path-info` **as a post-failure fallback, not a pre-flight gate.** If `buildenv.nix` fails, run `nix path-info` to check whether paths are missing from the Nix daemon's DB. If any are unregistered or absent, fetch/re-register them and try again. If all paths are confirmed present, run `buildenv.nix` once more — this second attempt is now preceded by a confirmed-clean `nix path-info`, covering the DB-registration-lag race.
3. **Replace** `MAX_RETRIES` **with a deterministic termination condition.** The correct loop invariant is not a retry count but a two-sided path confirmation: a failure is only deterministic when `nix path-info` returns valid results for all store paths **both before and after** a single `buildenv.nix` invocation. Concretely:
   * The post-failure `nix path-info` from one attempt serves as the pre-invocation confirmation for the next.
   * If `nix path-info` after a failure finds missing paths (GC raced during the call), those paths are re-materialised and the loop continues — regardless of how many times this has already happened.
   * Only when two consecutive `nix path-info` checks bracket a `buildenv.nix` failure with an all-paths-present result is the error treated as deterministic.

   In practice this loop terminates quickly: the GC-race-during-invocation scenario is rare, and each iteration resolves it. But a fixed retry count is the wrong mechanism — it would either under-count (terminate before the race resolves) or over-count (spin unnecessarily on a truly deterministic error).
4. **Remove** `is_transient()` **and** `BuildEnvError` **classification.** Retry eligibility is determined by path availability, not error type. The `is_transient()` method and any retry logic conditioned on error classification can be deleted.
5. **Remove backoff.** Backoff is only meaningful when you don't know if the thing you're waiting for is ready. With `nix path-info` confirming path state before each retry, we either fix the root cause and proceed immediately, or declare the failure deterministic. There is nothing to sleep through.

## Scope

* `cli/flox-rust-sdk/src/providers/buildenv.rs`: `materialise_with_retry`, `MAX_RETRIES`, `backoff_delay`, `is_transient`, `BuildEnvError` classification variants
* Remove or collapse the `materialise_with_retry` wrapper if the simplified logic no longer warrants a separate function
* Update or remove unit tests built around the retry/backoff/classification model

## References

* flox/flox#4282 — original `materialise_with_retry` introduction (merged)
* flox/flox#4327 — root cause identification (closed)
* flox/flox#4302, flox/flox#4303, flox/flox#4304, flox/flox#4305, flox/flox#4336 — retry/classification issues (closed as superseded)
* [#4323](<https://github.com/flox/flox/issues/4323>), [#4324](<https://github.com/flox/flox/issues/4324>) — retry/backoff/classification PRs (closed as superseded)

### PR DESCRIPTION
**Pull request by limeytexan**
Closes #4348.

## Summary

`materialise_with_retry` previously used `MAX_RETRIES = 3` and a fixed-count loop to handle Nix store DB lag and GC races during environment builds. This was based on a misdiagnosis of the root cause: `builtins.storePath` inside `buildenv.nix` reads from the Nix daemon's SQLite database, which can lag behind the filesystem. A retry count and sleep-based backoff are the wrong tools for this.

The new algorithm terminates on a deterministic condition rather than a count:

1. Call `realise` to fetch any missing store paths.
2. Stat-check all expected paths. If any are missing (GC raced with `realise`), restart.
3. Call `buildenv.nix` optimistically — no `nix path-info` overhead on the happy path.
4. On failure, run `nix path-info --json`:
   - **Unregistered paths found**: force-register / re-materialise and restart.
   - **All confirmed registered**: carry this clean result as the "before" check for the next iteration. Only when `nix path-info` is clean **both before and after** a single failed invocation is the error declared deterministic.
   - **`nix path-info` errors**: propagate the original error immediately.

The "both before and after" condition correctly handles the DB-registration-lag race (#4327) — a path can arrive on disk before the daemon records it in its DB, then registration completes before `nix path-info` runs — without requiring an arbitrary sleep or retry budget.

## Changes

- `BuildEnvError::MaterialisationFailed` removed (was only meaningful in the context of `MAX_RETRIES` exhaustion; `realise` errors still propagate immediately)
- `MAX_RETRIES` constant and `attempt` counter removed
- `materialise_with_retry` rewritten as an unbounded `loop` with `path_info_clean_before: bool` state
- Tests updated: `materialisation_failed_when_paths_always_missing_before_build_env` → `retries_until_gc_stops_before_build_env`; `materialisation_failed_when_gc_detected_after_build_env_on_final_attempt` removed (the new algorithm handles this without a cap); `build_env_error_exhausts_retries_when_paths_confirmed_present` → `build_env_error_declared_deterministic_after_two_clean_path_info_checks` (now asserts 2 `build_env` calls, not 3)

## Release Notes

No user-visible changes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 1]
```diff
@@ -1074,206 +1065,152 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
-                                warn!(
-                                    error = %e,
-                                    attempt,
-                                    MAX_RETRIES,
-                                    "buildenv.nix failed with all paths confirmed in Nix store — possible DB-registration race, retrying"
-                                );
-                                continue;
-                            }
+                let all_paths = expected_paths();
+                match nix_path_info_null_paths(&all_paths) {
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: The concern is unfounded. `nix path-info` exits 0 and returns `null` in the JSON for paths that are not in the Nix daemon's database — confirmed behaviour, regardless of whether the path was GC'd off disk or simply never registered. So a path that GC races away during `buildenv.nix` will appear as a null entry in the path-info output and land in `null_paths`.

That triggers the `Ok(null_paths)` branch, which calls `nix build --no-link --keep-going` to re-materialise / re-register the paths and then restarts from `realise()` with `path_info_clean_before = false`. The loop handles it correctly.

The `Err(path_info_err)` arm only fires when the `nix path-info` invocation itself fails (spawn error, malformed output, etc.) — a genuine hard error unrelated to whether the paths exist. Treating that as deterministic is correct; retrying would not help.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 2]
```diff
@@ -2818,52 +2753,67 @@ mod materialise_retry_tests {
...
+            "realise called on each loop iteration"
+        );
     }
 
-    // --- post-build GC detection (missing paths discovered after build_env fails) ---
+    // --- post-build GC detection (paths absent/unregistered after build_env fails) ---
 
     #[test]
     fn retries_when_gc_detected_after_build_env_failure() {
         init_tracing();
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: Agreed — the name is misleading. This test exercises the DB-registration-lag path: `build_env` fails once, `nix path-info` returns clean (no null paths), `path_info_clean_before` is set to `true`, the loop retries, and the second attempt succeeds. It never exercises the null-paths / force-register branch at all.

Will rename to `retries_on_db_registration_lag_after_build_env_failure` and update the comment block to match.

File: cli/flox-rust-sdk/src/providers/buildenv.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: Agreed. Will compute `expected_paths()` once per iteration before the `debug!` call and reuse the result for `nix_path_info_null_paths`. Avoids double evaluation and ensures the path list is consistent between logging and the path-info check.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 3]
```diff
@@ -1074,206 +1065,152 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
-                                continue;
-                            }
+                let all_paths = expected_paths();
+                match nix_path_info_null_paths(&all_paths) {
+                    Err(path_info_err) => {
+                        warn!(
+                            error = %e,
+                            path_info_error = %path_info_err,
+                            "buildenv.nix failed and nix path-info check errored \
+                            — treating as deterministic"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: Agreed. Per project convention, user-visible strings should stretch past the line-width limit rather than use `\` continuations (which embed the indentation whitespace). Will make this single-line.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 4]
```diff
@@ -1074,206 +1065,152 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
+                            // path-info was clean both before and after this
+                            // invocation: the failure is deterministic.
                             warn!(
                                 error = %e,
-                                attempt,
-                                MAX_RETRIES,
-                                "buildenv.nix failed with all paths confirmed in Nix store after all retries — treating as deterministic"
+                                "buildenv.nix failed with all paths confirmed in Nix \
+                                store both before and after the invocation \
+                                — treating as deterministic"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: Agreed. Per project convention, user-visible strings should stretch past the line-width limit rather than use `\` continuations (which embed the indentation whitespace). Will make this single-line.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 5]
```diff
@@ -1074,206 +1065,152 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
+                        // path-info is clean only "after" so far.  This covers
+                        // the DB-registration-lag race (#4327): a path can arrive
+                        // on disk (passing stat()) before the daemon records it in
+                        // its DB (causing builtins.storePath to fail), then the DB
+                        // registration completes before nix path-info runs.
+                        // Retry with this clean result as the "before" confirmation.
+                        warn!(
+                            error = %e,
+                            "buildenv.nix failed with all paths confirmed in Nix store \
+                            — possible DB-registration race, retrying"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: Agreed. Per project convention, user-visible strings should stretch past the line-width limit rather than use `\` continuations (which embed the indentation whitespace). Will make this single-line.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 6]
```diff
@@ -1074,206 +1065,152 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
+                    },
+                    Ok(null_paths) => {
+                        // Some paths are not registered in the Nix daemon's DB —
+                        // either present on disk but bypassing the daemon (e.g. a
+                        // CI cache that copies directly to /nix/store), or evicted
+                        // entirely.  Force-register / re-materialise and retry.
+                        warn!(
+                            ?null_paths,
+                            "buildenv.nix failed: paths not registered in Nix store \
+                            — force-registering and retrying"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: Agreed. Per project convention, user-visible strings should stretch past the line-width limit rather than use `\` continuations (which embed the indentation whitespace). Will make this single-line.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 7]
```diff
@@ -1074,206 +1065,152 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
                                 error = %e,
-                                path_info_error = %path_info_err,
-                                attempt,
-                                MAX_RETRIES,
-                                "buildenv.nix failed and nix path-info check errored \
-                                — treating as deterministic"
+                                status = %reg.status,
+                                stderr = %String::from_utf8_lossy(&reg.stderr),
+                                "failed to register unregistered paths \
+                                — propagating original error"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: Agreed. Per project convention, user-visible strings should stretch past the line-width limit rather than use `\` continuations (which embed the indentation whitespace). Will make this single-line.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 8]
```diff
@@ -1074,206 +1065,146 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
+    // came back clean.  A clean result at the end of iteration N is the
+    // "before" confirmation for the buildenv.nix call in iteration N+1.
+    // Only when both "before" and "after" checks are clean for the same
+    // invocation is the failure declared deterministic.
+    let mut path_info_clean_before = false;
+
+    loop {
         realise()?;
+
         let missing = missing_paths();
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs:1128]: Agreed. The scenario: iteration N sets `path_info_clean_before = true` (build failed, post-failure path-info clean); GC removes paths before iteration N+1; `realise()` re-materialises them; `missing_paths()` sees them all present; `build_env()` fails again due to DB-registration lag introduced by the fresh materialisation; post-failure path-info comes back clean; `path_info_clean_before` is still `true` → declares deterministic incorrectly.

Fix: call `missing_paths()` before `realise()`. If any are absent, reset `path_info_clean_before = false` before `realise()` runs — store state is about to change, so the prior path-info result is no longer a valid 'before' confirmation.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 9]
```diff
@@ -1074,206 +1065,149 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
+                    },
+                    Ok(null_paths) => {
+                        // Some paths are not registered in the Nix daemon's DB —
+                        // either present on disk but bypassing the daemon (e.g. a
+                        // CI cache that copies directly to /nix/store), or evicted
+                        // entirely.  Force-register / re-materialise and retry.
+                        warn!(
+                            ?null_paths,
+                            "buildenv.nix failed: paths not registered in Nix store — force-registering and retrying"
+                        );
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs:1194]: Agreed. `nix_path_info_null_paths` is called directly, so the `Err` arm and the non-empty `null_paths` branch (including the force-registration failure path) can't be exercised without a live Nix daemon. Will inject both `path_info` and `force_register` as closures alongside the existing parameters, and add unit tests for: path-info errors (propagate original error), null paths present + force-register succeeds (loop restarts), and null paths present + force-register fails (propagates original error).

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 10]
```diff
@@ -1074,206 +1065,141 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
+    // Only when both "before" and "after" checks are clean for the same
+    // invocation is the failure declared deterministic.
+    let mut path_info_clean_before = false;
+
+    loop {
+        // If paths are already absent before realise(), it will change store
+        // state; any prior path-info "before" confirmation no longer applies.
+        if !missing_paths().is_empty() {
+            path_info_clean_before = false;
+        }
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs]: Agreed — it's a no-op when the flag is already `false`. Will guard the check with `if path_info_clean_before && !missing_paths().is_empty()` so the extra stat sweep is skipped on every iteration where the flag isn't set.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 11]
```diff
@@ -1074,206 +1065,141 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
-            debug!(paths = ?missing, "full list of missing store paths after exhausting retries");
-            return Err(BuildEnvError::MaterialisationFailed {
-                attempts: MAX_RETRIES,
-                paths: display,
-            });
+            debug!(
+                ?missing,
+                "store paths missing after materialisation, GC suspected — retrying"
+            );
+            continue;
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs:1136]: In practice this scenario is pre-empted by `realise()`: if the Nix daemon has a path registered but it isn't visible on the local filesystem (broken NFS mount, FUSE layer not replicating), `realise()` — which runs `nix build` or `nix-store --realise` — will fail and propagate immediately via `?`, never reaching `missing_paths()`. The post-realise non-empty case therefore corresponds to GC removing a path between `realise()` and the stat check, which is exactly the race the loop is designed to handle. Persistent GC that never stops is bounded by the fact that a caller holding valid store path references prevents eviction. An explicit circuit breaker is worth a follow-up issue but is out of scope for this PR, which is specifically removing the arbitrary fixed count.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 12]
```diff
@@ -1362,6 +1286,26 @@ where
...
+                    .args(paths)
+                    .output()?;
+                if reg.status.success() {
+                    Ok(())
+                } else {
+                    Err(std::io::Error::other(format!(
+                        "exit {}: {}",
+                        reg.status,
+                        String::from_utf8_lossy(&reg.stderr).trim()
+                    )))
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs:1314]: Good catch. `ExitStatus::Display` already emits `exit status: N` (on Unix), so wrapping it in `"exit {}"` produces `"exit exit status: 1: ..."`. Will change the format string to `"{}: {}"` and let `ExitStatus` format itself.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 13]
```diff
@@ -1074,206 +1065,141 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
-                                warn!(
-                                    error = %e,
-                                    attempt,
-                                    MAX_RETRIES,
-                                    "buildenv.nix failed with all paths confirmed in Nix store — possible DB-registration race, retrying"
-                                );
-                                continue;
-                            }
+                match path_info(&all_paths) {
+                    Err(path_info_err) => {
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs:1153]: Agreed. When `all_paths` is empty, `path_info(&[])` always returns `Ok([])` regardless of store state (the short-circuit in `nix_path_info_null_paths` fires before anything is queried), so the two-sided check produces no signal and the loop retries once with a misleading 'DB-registration race' warning. Will add an early `return Err(e)` in the `build_env` failure arm when `all_paths.is_empty()`, skipping the path-info dance entirely.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 14]
```diff
@@ -1074,206 +1065,145 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
-                )
-            };
-            debug!(paths = ?missing, "full list of missing store paths after exhausting retries");
-            return Err(BuildEnvError::MaterialisationFailed {
-                attempts: MAX_RETRIES,
-                paths: display,
-            });
+            debug!(
+                ?missing,
+                "store paths missing after materialisation, GC suspected — retrying"
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs:1132]: Agreed. If `path_info_clean_before = true` and GC removes paths during `realise()`, the post-realise `missing_paths()` fires and the loop continues — but the flag stays set, carrying a stale 'before' confirmation into the next iteration. Will reset `path_info_clean_before = false` in this branch alongside the `continue`.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 15]
```diff
@@ -2872,76 +2835,109 @@ mod materialise_retry_tests {
...
-    fn materialisation_failed_when_gc_detected_after_build_env_on_final_attempt() {
+    fn build_env_error_declared_deterministic_after_two_clean_path_info_checks() {
         init_tracing();
-        // Paths always disappear after build_env fails — exhausts all retries.
-        let missing_call = Cell::new(0usize);
+        // build_env always fails; stat always shows paths present; path-info
+        // (empty expected_paths → Ok([])) is always clean.
+        // First failure: path_info_clean_before=false → sets it to true, retries.
+        // Second failure: path_info_clean_before=true → declares deterministic.
+        // Exactly two build_env calls and two realise calls.
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs:2917]: Good catch — the comment predates the change to non-empty `expected_paths`. Will update it to reflect that path-info is clean because the injected closure returns `Ok([])` unconditionally, not because the input is empty.

File: cli/flox-rust-sdk/src/providers/buildenv.rs [hunk 16]
```diff
@@ -1074,206 +1065,147 @@ fn nix_path_info_null_paths(paths: &[String]) -> Result<Vec<String>, std::io::Er
...
-                attempts: MAX_RETRIES,
-                paths: display,
-            });
+            debug!(
+                ?missing,
+                "store paths missing after materialisation, GC suspected — retrying"
+            );
+            // Store state changed; any prior path-info "before" confirmation is stale.
+            path_info_clean_before = false;
+            continue;
```
[the remark this answers is not in the document]
Comment (limeytexan) [reply cli/flox-rust-sdk/src/providers/buildenv.rs:1136]: Same concern as raised in an earlier comment (3363038411) — see the reply there for the full rationale. In short: `realise()` runs `nix build` or `nix-store --realise`, which will fail and propagate via `?` if it cannot make paths locally visible; it never returns `Ok(())` while leaving paths absent on the local filesystem in a working configuration. The post-realise non-empty case therefore corresponds exclusively to GC racing between `realise()` and the stat check — exactly the race this loop is designed to handle. A circuit breaker for truly degenerate cases (continuously racing GC) is a reasonable follow-up, but adding one here would reintroduce an arbitrary bound that this PR is specifically removing.

**Comment by limeytexan** [comment]
## Domain Summary

The materialisation retry loop changes from count-based to evidence-based
termination. Previously, the loop ran at most three times and aborted with
a `MaterialisationFailed` diagnostic if paths were still unavailable.
Now the loop continues until it can prove a failure is deterministic —
`nix path-info` must be clean both before and after a single `buildenv.nix`
invocation — or until a hard error occurs. On the success path, `nix path-info`
is no longer called at all (first call is now optimistic), which is a small
performance win on every healthy build. The `MaterialisationFailed` error
variant and `MAX_RETRIES` are removed entirely.

**Framing note from Lens B:** PRs #4323 and #4324 (backoff, transient-error
retry) were closed without merging; per issue #4348 that approach was a
misdiagnosis. This PR refactors the original plain `MAX_RETRIES=3` loop,
not a backoff-augmented one.

## Description Consistency

**Consistent:** The four-step algorithm, the "double-clean" determinism rule,
the enumerated removals (`MaterialisationFailed`, `MAX_RETRIES`, the `attempt`
counter), and the test renames all match the diff.

**Missing from description:** Two user-observable effects: (1) the success path
no longer invokes `nix path-info` (a performance improvement not called out);
(2) the GC-specific guidance message ("wait for garbage collection to finish and
retry") is gone — a user hitting GC contention now sees the raw `buildenv.nix`
error rather than a tailored explanation. "No user-visible changes" slightly
understates this.

## Code Review

### Critical

**[C] buildenv.rs:1126-1135 — unbounded loop when `realise()` succeeds but paths stay missing**

When `realise()` returns `Ok` but `missing_paths()` is perpetually non-empty —
sustained concurrent GC, NFS/FUSE visibility flapping, or a path that genuinely
never materialises — the loop does `path_info_clean_before = false; continue`
with no iteration cap and no termination bound. The old `MAX_RETRIES=3` loop
guaranteed termination and emitted `MaterialisationFailed`. The new code can
hang indefinitely instead. This is a real behavioral regression: the old code
failed loudly; the new code silently loops.

**[C] buildenv.rs:1184-1204 — unbounded loop when `force_register` exits 0 but paths remain unregistered**

`nix build --no-link --keep-going` exits 0 even when individual derivations
fail to build. If `force_register` returns `Ok(())` while the same `null_paths`
remain on the next `nix path-info` call, the loop resets `path_info_clean_before`
to false and continues. The two-sided "clean before and after" determinism check
is never reached, and the loop spins without bound. The removed test
`materialisation_failed_when_paths_always_missing_before_build_env` previously
locked in the "always missing → error" contract; nothing replaces that guarantee
in the new code.

Recommendation for both Critical findings: add a generous iteration or
elapsed-time backstop (e.g. after N iterations, return the original error with
a diagnostic noting the persistent store condition). This restores the guaranteed
termination the old code provided without reverting the evidence-based
correctness improvement.

### Important

**[I] buildenv.rs:1118-1122 — the core "two-sided invalidation" branch is never tested**

Lines 1118-1122 are the namesake behavior: they reset a carried-forward
`path_info_clean_before` confirmation when a stat-level path disappearance
occurs between two failing iterations, preventing a stale "clean" reading from
marking a GC-interrupted failure as deterministic. Every existing test keeps
`path_info_clean_before == false` during this check — the test comments confirm
the branch is always skipped. A test is needed where: build_env fails with clean
path-info (sets flag true), a path then goes missing before the next realise
(must reset flag), build_env fails again with clean path-info — and the result is
a *retry*, not a deterministic exit. Without it, a regression in the invalidation
logic passes CI.

**[I] buildenv.rs:1184 — force-register resets the two-sided flag, interaction with finding [C2] undocumented**

When force-register succeeds, `path_info_clean_before` is set to false and the
loop continues. This is defensible (store state changed), but it means an
environment whose paths are repeatedly evicted and re-registered can never
accumulate the two consecutive clean checks needed to declare determinism. Worth
a comment acknowledging the interaction with the termination guarantee.

**[I] Empty-path build failures now surface on the first attempt rather than after three**

`if all_paths.is_empty() { return Err(e); }` returns immediately on the first
`build_env` failure when there are no store paths. Previously an empty-path
deterministic failure took 3 attempts. The current behavior is correct and
desirable — confirm it is intentional. If any empty-path *transient* failure
exists, it now gets zero retries.

**[I] buildenv.rs:1120 — `missing_paths()` called twice per iteration on the retry path**

The pre-realise guard adds a second `missing_paths()` call (on top of the
post-realise call at 1126) when `path_info_clean_before` is true. On large
environments this is O(paths) `stat()` syscalls twice per retry iteration. The
`&&` short-circuit ordering keeps this off the success path — that ordering is
load-bearing and worth a comment so a future refactor doesn't accidentally double
the stat cost on every build.

### Minor

**[M] buildenv.rs:1085-1092 — doc comment overstates the "before" path-info check**

The rustdoc says `nix path-info` is clean "both before and after a single
`buildenv.nix` invocation." In reality, "before" is the previous iteration's
`path-info` result carried forward in `path_info_clean_before`, re-validated
before the next invocation only by a `stat`-level `missing_paths()` check —
not by re-running `nix path-info`. The inline comment at lines 1110-1114
describes this accurately; the rustdoc should match.

**[M] Residual "attempt" vocabulary in test names**

`succeeds_on_first_attempt_when_paths_present` and assertion messages like
`"realise called on each attempt"` retain the old attempt-counter vocabulary.
The surrounding new comments use "iteration." Minor inconsistency worth
aligning.

---
*Via Forge (interactive) • 1b4da53a*

**Comment by limeytexan** [comment]
Withdrawing [C1]. The "hangs indefinitely" framing was wrong. When NFS is unresponsive, the underlying stat() and nix subprocess calls block at the kernel level — an arbitrary count-based ceiling doesn't rescue from that, it just picks an arbitrary point to return a misleading error. The software can't know whether the filesystem condition will clear; surfacing the inconsistency and letting the operator decide is the correct design. The loop rate-limits itself naturally since each iteration involves real subprocess work, and the operator can kill the process if they conclude the condition won't resolve.

[C2] is a narrower case — nix itself reporting success via `--keep-going` while a specific path remains unregistered is a misleading signal from nix, not a filesystem service being down. But each iteration still involves real subprocess work, so it's not a tight spin. The concern is better framed as a logging gap: a warn! on each retry iteration would surface what's happening to the operator without imposing an arbitrary termination count. That aligns with the "proactively report the inconsistency" design intent.

Remaining items:
- The "two-sided invalidation" branch (lines 1118-1122) being untested is still the sharpest gap
- The doc comment overstating the "before" path-info check as a re-run rather than a carried-forward result
- Per-iteration warn!/debug! on the retry branches so operators can observe what the loop is doing

---
*Via Forge (interactive) • 1b4da53a*

**Comment by limeytexan** [comment]
Revising the C1 withdrawal — the termination behavior is correct, but there is still a real gap: observability.

**C1 and C2 are distinct code paths, both currently silent:**

- **C1** (top of loop): `realise()` returns `Ok` but `missing_paths()` shows paths still absent on disk. This is the faulty-remote-filesystem / NFS-lag scenario — nix reports the path was written, but the client can't see it. The code resets `path_info_clean_before` and `continue`s with no log output. An operator whose build is stuck here has no signal equivalent to NFS's "server not responding... retrying".

- **C2** (force-register branch): `force_register()` returns `Ok` but the next `path-info` still shows null paths — nix DB inconsistency. Also currently silent if this repeats.

**What's needed:** a `warn!` on each C1 and C2 iteration so operators can observe what the loop is doing, with decreasing log frequency per successive iteration (e.g. every N^th repetition) to avoid spamming logs on a sustained condition while still making the problem visible.

The NFS analogy is exact for C1: clients emit "NFS server not responding... retrying" periodically rather than every syscall, and that's the right model here — surface that something is wrong and that the system is actively retrying, without burying the signal in noise.

Could you confirm whether the current implementation emits anything on either of these branches, and if not, add periodic warn! messages before merge?

---
*Via Forge (interactive) • 1b4da53a*

**Comment by limeytexan** [comment]
Thanks for the thorough review. Addressing each item:

**C1 / C2 — observability**

- **C1 (post-realise missing paths):** the branch was emitting `debug!`, invisible to operators by default. Changed to `warn!` with the missing-path list as a structured field. Each iteration involves a real `nix build` subprocess, so log volume is naturally bounded; a separate rate-limiting counter would add complexity without much benefit.
- **C2 (force-register → null paths on next path-info):** already emits `warn!(?null_paths, "buildenv.nix failed: paths not registered in Nix store — force-registering and retrying")` at the top of each null-paths branch pass, so operators see the repeating condition.

**Untested two-sided invalidation branch (lines 1118-1122)**

Added `path_info_clean_before_reset_when_paths_absent_before_realise`. The missing-paths closure returns non-empty on even-numbered calls (pre-realise checks, which only fire when the flag is true) and empty on odd-numbered calls (post-realise checks). Build-env fails on iterations 1 and 2 with clean path-info; without the reset, the second clean result would declare the failure deterministic after 2 build calls. With the reset, the loop correctly reaches a third call that succeeds. Removing the invalidation branch causes this test to fail.

**Doc comment — Termination paragraph**

Updated from "when `nix path-info` is clean both before and after a failed invocation" to make explicit that the "before" is the *carried-forward result from the previous iteration*, not a fresh re-run around the current invocation.

**Vocabulary**

Renamed `succeeds_on_first_attempt_when_paths_present` → `succeeds_on_first_iteration_when_paths_present` and updated assertion messages to use "iteration" throughout.

**Comment by limeytexan** [comment]
All items resolved:

- **C1 observability**: `debug!` → `warn!` with structured missing-path field. The "each iteration involves a real nix subprocess" argument against rate-limiting is sound — log volume is naturally bounded by subprocess cost.
- **C2 observability**: already emitted `warn!` — good to know the branch wasn't silent, just not visible in the diff review.
- **Untested two-sided invalidation**: new test `path_info_clean_before_reset_when_paths_absent_before_realise` sounds exactly right, and the "removing the branch causes the test to fail" confirmation is the key verification.
- **Doc comment** and **vocabulary**: both corrected.

PR is clear to merge.

---
*Via Forge (interactive) • 1b4da53a*
